python-frontend/src/main/java/org/sonar/python/types/v2/TypeCheckBuilder.java

```diff
@@ -139,7 +139,7 @@ public TriBool test(PythonType pythonType) {
       if (pythonType instanceof ObjectType objectType) {
         pythonType = objectType.unwrappedType();
       }
-      if (pythonType == PythonType.UNKNOWN || expectedType == PythonType.UNKNOWN) {
+      if (pythonType instanceof UnknownType || expectedType instanceof UnknownType) {
```

Comment:
I would expect to see a test that breaks without this change.

As I mentioned in my first comment, my concern with the initial implementation was that we might break some logic when we assume `expectedType == PythonType.UNKNOWN` is enough to test that a type is unknown (since it no longer is the case).

In a TDD approach I think it would have made more sense to create such a test to prove the concern right before changing the implementation.

Reply:
I have added the test.
I tried to add tests, but I don't think some scenarios are reachable with the current type inference.